### PR TITLE
Issue #3131: Add support for try with resources

### PR DESCRIPTION
**Pull request by kukreja-vikramaditya**
Issue: #3131

### CODE REVIEW

**Comment by rnveach** [comment]
@kukreja-vikramaditya Werker failure is due to regression issue on orekit.
> [ERROR] src/main/java/org/orekit/files/sp3/SP3Parser.java:[190] (indentation) Indentation: 'try' have incorrect indentation level 13, expected level should be 12.
[ERROR] src/main/java/org/orekit/files/sp3/SP3Parser.java:[191] (indentation) Indentation: 'try' have incorrect indentation level 13, expected level should be 12.

I will look over the PR and orekit's files. If requested changes seem valid, we will have to make a PR with them to correct regression.

**Comment by kukreja-vikramaditya** [comment]
Reports at https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/index.html
There seems to be false positives in many of the projects. Eg. in cases with a '\t' instead of spaces.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputMultilineTryWithResources.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+        try ( //indent:8 exp:8
+BufferedWriter writer = Files.newBufferedWriter(outputFilePath, charset); //indent:0 exp:12 warn
+            ZipFile zf = new ZipFile(zipFileName) //indent:12 exp:12
+        ) { //indent:8 exp:8
+            zf.getName(); //indent:12 exp:12
+        } //indent:8 exp:8
+        try ( //indent:8 exp:8
+            BufferedWriter writer = Files. //indent:12 exp:12
+newBufferedWriter(outputFilePath, charset); //indent:0 exp:0
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputMultilineTryWithResources.java]: no warning here is definitely wrong.
Can we fix this in scope of this PR?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputMultilineTryWithResources.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    } //indent:4 exp:4
+
+    static void fooMethod(String zipFileName) throws IOException { //indent:4 exp:4
+
+        Charset charset = StandardCharsets.US_ASCII; //indent:8 exp:8
+        Path outputFilePath = Paths.get(zipFileName); //indent:8 exp:8
+
+        try ( //indent:8 exp:8
+BufferedWriter writer = Files.newBufferedWriter(outputFilePath, charset); //indent:0 exp:12 warn
+            ZipFile zf = new ZipFile(zipFileName) //indent:12 exp:12
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputMultilineTryWithResources.java]: I don't see any examples where 2nd variable is lower or higher than expected.
Please add them.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputValidTryResourcesIndent.java [hunk 3]
```diff
@@ -26,7 +26,7 @@
     private static Manifest loadManifest(String fn) { //indent:4 exp:4
         try (FileInputStream fis = new FileInputStream(fn); //indent:8 exp:8
     // This should be an error //indent:4 exp:4
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputValidTryResourcesIndent.java]: remove this comment since it is fixed now.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 4]
```diff
@@ -42,10 +43,31 @@ public TryHandler(IndentationCheck indentCheck,
 
     @Override
     public IndentLevel getSuggestedChildIndent(AbstractExpressionHandler child) {
+
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: remove unnecessary changes.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 5]
```diff
@@ -42,10 +43,31 @@ public TryHandler(IndentationCheck indentCheck,
...
         if (child instanceof CatchHandler
             || child instanceof FinallyHandler) {
             return getIndent();
         }
+
+        final DetailAST resSpecAst = getMainAst().getFirstChild();
+        if (resSpecAst.getType() == TokenTypes.RESOURCE_SPECIFICATION) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: This is a weird place to add the code.
This method is getting expected child indent, but you are doing validation on non-children elements and not adding to how the expected child indent should be calculated.

I think we should move this code to a new method that overrides `checkIndentation`, still calls the super, and then does your extra validation. This is how `MethodDefHandler` does it too.
This sound more reasonable?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 6]
```diff
@@ -42,10 +43,31 @@ public TryHandler(IndentationCheck indentCheck,
...
+            final IndentLevel expIndent = new IndentLevel(getIndent(), getBasicOffset());
+            DetailAST resourcesAstChild = resourcesAst.findFirstToken(TokenTypes.RESOURCE);
+
+            for (int i = 0; i < resourcesAst.getChildCount(); i++) {
+
+                if (resourcesAstChild.getType() == TokenTypes.RESOURCE
+                    && resourcesAstChild.getLineNo() != resSpecAst.getLineNo()
+                    && resourcesAstChild.getColumnNo()
+                        != Integer.parseInt(expIndent.toString())) {
+                    logError(resourcesAstChild, "",
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: This probably relates directly to my comment about the no warning in the tests.
You are just making sure where the resource starts on the line, and not all the lines of the resource.
If we call the line wrapping handler here for the entire 1 resource, it would handle the wrapping for us.
What do you think of this?

**Comment by rnveach** [comment]
> here seems to be false positives in many of the projects.

Indentation also requires finesse in regression as it is entirely dependent on the indenting the project uses. If you don't use the correct indentation that the projects require, it will make the report look weird and harder to understand.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java (hunk 6 above)
Comment (kukreja-vikramaditya) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: line wrapping handler for the resource ast is a good idea. I earlier tried resolving this using checkExprSubtree, but to no avail.

**Comment by kukreja-vikramaditya** [comment]
@rnveach 
Can I go with the assumption that all try with resources would be of the form


**Case 1: (*good*)**
```java
//General case with ")"
BufferedWriter writer = Files.newBufferedWriter(outputFilePath, charset)
```
where there is always a **`)`** or should I still consider these cases :

**Case 2: (*bad?*)**
```java
//Initialise outside try block
BufferedWriter writ = Files.newBufferedWriter(outputFilePath, charset);
try (BufferedWriter writer = writ) {
            //...
        } 
        ...
```
and this too?

**Case 3: (*bad?*)**
```java
//from a static object reference
try (BufferedWriter writer = MyClass.staticRefToWriterObject) {
            //...   
        } 
        //...
```

**Comment by rnveach** [comment]
@kukreja-vikramaditya If we go with an assumption code will always be written a specific way, that will probably lead to weird indentation being requested.
The check should only require that we show them how to indent everything correctly, not that they should rewrite their code. We should be able to indent anything successfully, no matter how it is written.

In terms of this PR, we should try to fix this as best as we can. Indentation has alot of weird issues and nothing will be perfect for now.
Anything that is still broken, can probably be put into a new issue if no one sees a way to fix it.

**Comment by rnveach** [comment]
@kukreja-vikramaditya 
I know the report for hbase is very big, but please look into it as there is an exception in the difference.
https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/Hbase/index.html#A46
https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/Hbase/xref/home/vikramaditya/Desktop/CS/contribution-master/checkstyle-tester/repositories/Hbase/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestAccessController.java.html#L0

>  	Checker 	Got an exception - java.lang.NumberFormatException: For input string: "16, 20"

**Comment by kukreja-vikramaditya** [comment]
@rnveach 
I have updated the PR and diff reports. But there is still regression. I'll go thru the Hbase report also as it has the maximum warnings. Could you please go thru the PR and a few reports and finalise upon what should be the indentation levels? For normal cases - 12 or 16 and for line wrapping cases 20?

**Comment by kukreja-vikramaditya** [comment]
The `NumberFormatException` is still not clear to me. The link provided is always at line 0

**Comment by rnveach** [comment]
> The NumberFormatException is still not clear to me. The link provided is always at line 0

@kukreja-vikramaditya Why don't you look at your code and look for something that can throw a `NumberFormatException`. You can also manually run your PR code on that file to try and debug it.

**Comment by kukreja-vikramaditya** [comment]
I realised my mistake and this time I used `getIndent().getFirstIndentLevel()`.

 Upon generating the report for Hbase only, I got around 265 warnings. But I am a little doubtful about what should be indentation, 12?16?. While going thru the Hbase report I came across cases where `try` was at an offset of only 2 from previous line which itself was at an indentation of 2 (So value of `getIndent()` would not be equal to `tryAst.getColumnNo()` ) Because of this their children suffered. Whether they should follow `getIndent() + offset` or `tryAst.columnNo() + offset`

Most of the examples in the report are like this except a few which were at odd positions and seem be correct detection. 

Updated report for Hbase at https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/Hbase-for-4171/index.html

**Comment by rnveach** [comment]
> I realised my mistake and this time I used getIndent().getFirstIndentLevel().

Why `FirstIndentLevel`? If multiple indentations are acceptable, then we should allow any one of them.
We should probably use `isAcceptable` to check indentation level against multiple allowed indentations.
Please also update your PR with your latest code.

> But I am a little doubtful about what should be indentation, 12?16?.

Can you give me a specific example with config, preferably from your tests. According to issue review, it should require base position + `lineWrappingIndentation`. Can you tell me where it is getting the other acceptable indentation level from?

> Whether they should follow getIndent() + offset or tryAst.columnNo() + offset

I have seen some areas in Indentation use one or the other. When everything is indented correctly, either one doesn't matter.
I think we should use `getIndent() + offset` so people don't think the children are indented correctly if the parent is wrong.
@romani Are you ok with this?

> OR maybe Hbase uses basicOffset of 2 

This is my configuration for hbase:
````
throwsIndent: 2
arrayInitIndent: 2
caseIndent: 2
basicOffset: 2
lineWrappingIndentation: 2
````

**Comment by kukreja-vikramaditya** [comment]
I am currently working on this sample input file. Please advice whether or not the cases have been handled properly http://pastebin.ubuntu.com/24360707/

**Comment by rnveach** [comment]
@kukreja-vikramaditya 
> 38 - newBufferedWriter(outputFilePath, charset); //indent:0 exp:17 warn

where did 17 come from when previous line is 8? I assume you are using values of 4 in config.
Shouldn't expected be 12 or 16?

I didn't see any others that stood out right now.
Please also add example you made from issue where 2nd resource and wrapping is wrong.

````
                ZipFile zf = new ZipFile( //indent:15 exp:16 warn
                   zipFileName) //indent:19 exp:19? or 20?  warn?
````

**Comment by kukreja-vikramaditya** [comment]
@rnveach 
@romani 

I have updated my PR with the changes. Reports have been updated [here](https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/index.html) (all except `Hbase` (will update after changing config) and `openjdk7` (groovy out of memory error)). 

Here is my findings, for the 11 repos facing warnings I have only checked 9 as of now,

`findbugs` and `find-bugs` are okay
`orekit` is okay
`elasticsearch` is okay
`checkstyle` and `checkstyle-excludes` are okay except for a line#27 in one of the input files
`jool` is okay
`apache-struts` is okay

`pmd` **errored** - it has `\t` which is causing issue. (Need to resolve this)

I'll check `openjdk8` and `openjdk9` soon and update. One line I noticed for `openjdk8` #2file line#834, `rparen` of `try` is showing wrong indentation as exp but `try` isn't.. which mean I must have missed check for `try`.

**Comment by rnveach** [comment]
> groovy out of memory error

Make sure you give groovy atleast 4 gigs of memory for this check.
This error can also mean indentation configuration is wrong for project and too many violations are being produced.

You mentioned problems with the reports, I will wait on those being fixed.

**Comment by kukreja-vikramaditya** [comment]
From the discussion at comment https://github.com/checkstyle/checkstyle/issues/3131#issuecomment-293257922 after your **EDIT** message, resources should follow the indentation of the first resource, what should be the indentations in the following case- 
```java
    try (                       BufferedWriter writer = ... 
         BufferedWriter writ = .. //indent:9 exp: 8? 9? 10? or directly below previous resource?
    ) {
```

Similar ques if first resource is not on the first line, what should be its indent (and other resources indent) `getIndent() + getBasicOffset()`?

This reason for this question is because mostly all warnings in the reports are of such cases ( whether it should be 13 or 12? 17 or 16? and so on.

>One line I noticed for openjdk8 here, rparen of try is showing wrong indentation as exp but try isn't.. which mean I must have missed check for try.

It wasn't `try` which was wrong but `rparen` indentation that was misleading. Still don't know why. Is my calling of the `checkRightParen` wrong?

**Comment by rnveach** [comment]
> what should be the indentations in the following case-

**Comment by rnveach** [comment]
> what should be the indentations in the following case-

The following code should be requested to be indented the same way:
````
    try (                       BufferedWriter writer = ... 
         BufferedWriter writ = .. //indent:9 exp: 8? 9? 10? or directly below previous resource?
    ) {
    try (
         BufferedWriter writer = ... 
         BufferedWriter writ = ..
    ) {
````
It shouldn't matter where `writer` is, all try resources that are first on line should be requested to be indented in the same position.

> getIndent() + getBasicOffset()?

See my post at https://github.com/checkstyle/checkstyle/issues/3131#issuecomment-293030061 .
> The basic concensus is, it looks like all IDEs treat the resources as wrapped lines ...
> offset of a resource should be try position + lineWrappingIndentation.
offset of wrapped resource should be resource position + lineWrappingIndentation.

> Is my calling of the checkRightParen wrong?

https://github.com/checkstyle/checkstyle/pull/4171/files#diff-ba5178dcaec15d23d5fa9819751dac1aR112
The call looks right. Left parenthesis and right parenthesis should be at the same position as long as right parenthesis starts the line.

**Comment by kukreja-vikramaditya** [comment]
@rnveach 
This PR has been hanging for some time so I tried simplifying the existing code, include checking only for resources that are first on line, used line wrapping checking for resources etc. 
Reports have been updated here 

Support for Line 1 has is still not added as I think it can be the job of whitespace checks to log errors for this? For resources that start new line however, expected indent is
`getIndent() + linewrappingIndent`. This however mean if line wrapping is 4, resources are not lined up like Intellij does. So my question is, should we manually specify a value of `5` or `6` from getIndent() or we should include some property that user can customise?

**Comment by romani** [comment]
@rnveach , what is a status of this PR ?
We do not need hanging PRs if they are in bad and unlikely to be successful.

**Comment by kukreja-vikramaditya** [comment]
@romani 
Will complete this

@rnveach 
Please see my previous comment

**Comment by rnveach** [comment]
> Support for Line 1 has is still not added as I think it can be the job of whitespace checks to log errors for this (Any thoughts on this)? 

Only the `try` should be validated by this check. Anything else on the line should be validated by some other check and not indentation. Indentation only cares about the first thing on the line.

> This however mean if line wrapping is 4, resources are not lined up like Intellij does.
> So my question is, should we manually specify a value of 5 or 6 from getIndent() to ensure lining up of resources, or we should include some property that user can customise?

@romani Please confirm if we should have special processing just for IntelliJ. See https://github.com/checkstyle/checkstyle/issues/3131#issuecomment-293257922 . Orekit seems to follow this pattern, so to pass regression I think we will have to do it.
Should it be embedded into the check to accept it, or should it be a special parameter, or ...?

> Reports have been updated here https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/index.html

Almost all violations are now new, and most want resources inside try to be double the set lineWrap indentation.

https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/openjdk8/index.html#A1
>  	'try' has incorrect indentation level 13, expected level should be 12.

Can we make the violation say 'try resource' instead of `try`? It should specify the area it is for and not it's parent.

**Comment by kukreja-vikramaditya** [comment]
@rnveach 
I have rebased on latest master and updated error message to include "resource" when resources start line
>Only the try should be validated by this check. Anything else on the line should be validated by some other check and not indentation. Indentation only cares about the first thing on the line.
Nothing to worry about in this PR right?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 7]
```diff
@@ -41,11 +42,71 @@ public TryHandler(IndentationCheck indentCheck,
...
+        }
+    }
+
+    /**
+     * Checks indentation of resources parameters in try block.
+     *
+     * @param resSpecAst Resource specification ast
+     */
+    private void checkTryResources(final DetailAST resSpecAst) {
+        final DetailAST ressAst = resSpecAst.findFirstToken(TokenTypes.RESOURCES);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: Use full words.
`ress` => `resources`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 8]
```diff
@@ -41,11 +42,71 @@ public TryHandler(IndentationCheck indentCheck,
...
+        final IndentLevel expResIndent = new IndentLevel(getIndent(),
+                getIndentCheck().getLineWrappingIndentation());
+        DetailAST resAst = ressAst.getFirstChild();
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: `res` => `resource`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 9]
```diff
@@ -41,11 +42,71 @@ public TryHandler(IndentationCheck indentCheck,
...
+    }
+
+    /**
+     * Checks indentation of resources parameters in try block.
+     *
+     * @param resSpecAst Resource specification ast
+     */
+    private void checkTryResources(final DetailAST resSpecAst) {
+        final DetailAST ressAst = resSpecAst.findFirstToken(TokenTypes.RESOURCES);
+        final IndentLevel expResIndent = new IndentLevel(getIndent(),
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: `expRes` => `expectedResource`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 10]
```diff
@@ -41,11 +42,71 @@ public TryHandler(IndentationCheck indentCheck,
...
+            checkTryResources(getMainAst().getFirstChild());
+        }
+    }
+
+    /**
+     * Checks indentation of resources parameters in try block.
+     *
+     * @param resSpecAst Resource specification ast
+     */
+    private void checkTryResources(final DetailAST resSpecAst) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: `resSpec` => `resourceSpec`

**Comment by rnveach** [comment]
@romani Are we going to try and finish this or are we going to wait until end of GSoC?

**Comment by romani** [comment]
@kukreja-vikramaditya, sorry for delay.

I relaunched Travis and TC.

> Please confirm if we should have special processing just for IntelliJ

it is hard to conflicting with IDEs, user might need to use `forceStrictCondition=false` ([doc](http://checkstyle.sourceforge.net/config_misc.html#Indentation )) it is  `false` by default.
it is default in Orekit - https://github.com/CS-SI/Orekit/blob/master/checkstyle.xml#L32.

from Wercker:
> 
```
[INFO] There are 2 errors reported by Checkstyle 8.0-SNAPSHOT with 
/pipeline/source/Orekit/checkstyle.xml ruleset.
[ERROR] .../SP3Parser.java:[190] (indentation) Indentation: 
'try resource' has incorrect indentation level 13, expected level should be 12.
[ERROR] .../SP3Parser.java:[191] (indentation) Indentation: 
'try resource' has incorrect indentation level 13, expected level should be 12.
```
[source lines](https://github.com/CS-SI/Orekit/blob/master/src/main/java/org/orekit/files/sp3/SP3Parser.java#L190)
so there should be no violation as it should be allowed any indentation that it bigger than 12.

**Comment by rnveach** [comment]
@kukreja-vikramaditya Please also rebase on latest master.

**Comment by romani** [comment]
looks like rebased but there still problems with regression, from wercker:
```
[INFO] 
[INFO] --- maven-checkstyle-plugin:2.17:check (default-cli) @ orekit ---
[INFO] There are 2 errors reported by Checkstyle 8.0-SNAPSHOT with /pipeline/source/Orekit
/checkstyle.xml ruleset.
[ERROR] src/main/java/org/orekit/files/sp3/SP3Parser.java:[190] (indentation) Indentation: 'try
 resource' has incorrect indentation level 13, expected level should be 12.
[ERROR] src/main/java/org/orekit/files/sp3/SP3Parser.java:[191] (indentation) Indentation: 'try
 resource' has incorrect indentation level 13, expected level should be 12.
```

@kukreja-vikramaditya , please continue with fix, let us know if we can help.

**Comment by kukreja-vikramaditya** [comment]
@romani @rnveach 
I believe I have solved the orekit issue as i am not facing that problem locally.
Could you pls share error log with me?

**Comment by rnveach** [comment]
From wercker:
> [INFO] Starting audit...
[ERROR] /pipeline/source/htmlunit/src/test/java/com/gargoylesoftware/htmlunit/protocol/data/DataURLDecoder2Test.java:91: 'try rparen' has incorrect indentation level 16, expected level should be 8. [Indentation]
Audit done.

**Comment by kukreja-vikramaditya** [comment]
@rnveach http://htmlunit.sourceforge.net/xref-test/com/gargoylesoftware/htmlunit/protocol/data/DataURLDecoder2Test.html#L91

It seems to me that the warning is correct. Any thoughts on this?

**Comment by Vladlis** [comment]
@kukreja-vikramaditya 
Please send a PR with fix to Orekit. If it is not accepted soon we will switch to fork.

**Comment by romani** [comment]
> It seems to me that the warning is correct. Any thoughts on this?

code is definitely could be written better. But why is is there a warning ? 
line was wrapped, indentation is the same as lines above. 
Fact that there are no statement there should not bother Check. Where to put wrapping is not business of indentation Check.
I think it is false-positive.

**Comment by rnveach** [comment]
> Where to put wrapping is not business of indentation Check.

@romani It is not specifying where to place wrapping. It is specifying where to indent wrapping.
Code is currently like:
````
90          try (ByteArrayInputStream in = new ByteArrayInputStream(decoder.getBytes());
91                  ) {
````
Indentation is requesting it be like:
````
90          try (ByteArrayInputStream in = new ByteArrayInputStream(decoder.getBytes());
91          ) {
````

> It seems to me that the warning is correct

@kukreja-vikramaditya Look at your examples from the major IDEs at https://github.com/checkstyle/checkstyle/issues/3131#issuecomment-292514375 .

Netbeans follows the indentation of htmlunit while IDEA and Eclipse don't. So this is a false positive and you must accept indentation of either one.
Please fix your code. No PR to htmlunit should be done.

**Comment by romani** [comment]
@kukreja-vikramaditya , please squash all your commits in one and rebase on latest code, to avoid conflicts.

**Comment by romani** [comment]
All CIs passed, that is good sign.
@rnveach , please finish your code review.
@kukreja-vikramaditya , please launch regression testing and share report.

File: config/checkstyle_checks.xml [hunk 11]
```diff
@@ -122,6 +122,7 @@
 
   <!-- Size Violations -->
   <module name="FileLength">
+    <property name="max" value="2500"/>
```
Comment (rnveach) [inline config/checkstyle_checks.xml]: Why did we change the configuration? This is not expected.

Comment (kukreja-vikramaditya) [inline config/checkstyle_checks.xml]: The tests I added resulted in increase of code lines greater than 2000 in the IndentationCheckTest file.
Should I instead use suppression?

**Comment by kukreja-vikramaditya** [comment]
@rnveach 
@romani 
Reports are here https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/
(I will add 4 more repositories result after downloading them)
Hbase has many errors but mostly because of indent value..will fix with remaining 4 repos

Reports look positive to me..I have also included a link of Old reports for this PR in case you wish to know the changes that happened after improving code

File: config/checkstyle_checks.xml (hunk 11 above)
Comment (rnveach) [inline config/checkstyle_checks.xml]: Suppressions in tests is fine for me. Indentation is a monster anyways.

**Comment by romani** [comment]
@kukreja-vikramaditya ,

http://checkstyle.sourceforge.net/config_misc.html#Indentation , default for `lineWrappingIndentation` is `4`.
Test case https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/infinispan/index.html#A1
please explain why indentation as 16 is expected, existing code looks valid for me as next line from `try` are intended on >4  spaces.

**Comment by kukreja-vikramaditya** [comment]
Yes it seems that I skipped the indentation case of Netbeans(where the case you mentioned has indentation, `lineWrappingIdent + try pos` .
IDEA and eclipse indents such cases by 4 (that is why reports showed 16 as valid)
So I shall add this as another acceptable indentation right?

**Comment by romani** [comment]
> Will fix this is next push soon and regenerate reports

please keep adding such cases to UTs.

**Comment by kukreja-vikramaditya** [comment]
@rnveach 
@romani 
Please check updated reports here https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/

**Comment by kukreja-vikramaditya** [comment]
From my code,
https://github.com/checkstyle/checkstyle/pull/4171/files#diff-ba5178dcaec15d23d5fa9819751dac1aR151

A check is done for wrapped resource, where the starting indentation is `try + wrap` and on top of that `wrap` is added . So it suggests using 16 instead of 12.

Should I specifically check for resource (that is in line with try) and use the `try + wrap` pos or should I use the current general case of `try + wrap + wrap` ?

**Comment by romani** [comment]
> Should I specifically check for wrapped resource (that is in line with try) and use the try + wrap pos

I think this is expected behavior. To make sure, lets recheck IDEs behavior, as how to wrap already discussed number of times before this day and already implemented in IDEs and user follow that  rules
If user need to indent to the right more he can do this (if strict mode is off).

@rnveach , please share your ideas.

**Comment by kukreja-vikramaditya** [comment]
@romani @rnveach 

All CIs are green :green_heart::green_heart::green_heart: .
 Reports on latest code here https://kukreja-vikramaditya.github.io/checkstyle-diff-reports/fix-issue-3131/diff/
Any suggestion for further improvement?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 12]
```diff
@@ -48,4 +67,94 @@ public IndentLevel getSuggestedChildIndent(AbstractExpressionHandler child) {
...
+     *
+     * @param parenAst      lparen or rparen ast to check
+     * @param subType       name to be used in log message
+     */
+    protected void checkTryResParen(final DetailAST parenAst,
+                                    final String subType) {
+        if (isOnStartOfLine(parenAst)) {
+            final IndentLevel expectedIdent = new IndentLevel(getIndent(),
+                    0,
+                    getIndentCheck().getLineWrappingIndentation());
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: please make it like:
```
            final IndentLevel expectedIdent = new IndentLevel(getIndent(), 0,
                    getIndentCheck().getLineWrappingIndentation());
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java [hunk 13]
```diff
@@ -48,4 +67,94 @@ public IndentLevel getSuggestedChildIndent(AbstractExpressionHandler child) {
...
+
+    /**
+     * Checks indentation of resources parameters in try resources.
+     *
+     * @param resourcesSpecAst   Resource specification ast
+     */
+    protected void checkTryResources(final DetailAST resourcesSpecAst) {
+        final DetailAST resourcesAst = resourcesSpecAst.findFirstToken(TokenTypes.RESOURCES);
+        final IndentLevel expectedResourceIndent = new IndentLevel(getIndent().getFirstIndentLevel()
+            + getIndentCheck().getLineWrappingIndentation());
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]:
```
final String indetation = getIndent().getFirstIndentLevel() + getIndentCheck().getLineWrappingIndentation()
final IndentLevel expectedResourceIndent = new IndentLevel(indetation);
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java (hunk 12 above)
Comment (kukreja-vikramaditya) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: I just noticed this too, is 8 fine? isn't 4 the default line wrap

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java (hunk 13 above)
Comment (kukreja-vikramaditya) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/TryHandler.java]: The first still needs to be broken.

**Comment by kukreja-vikramaditya** [comment]
Travis failure seems unrelated to the issue

**Comment by romani** [comment]
@kukreja-vikramaditya , please rebase one more time, I reverted my commit, that does not work for all.

**Comment by rnveach** [comment]
@kukreja-vikramaditya Please rebase one more time as latest master changed `createCheckConfig` to `createModuleConfig`. You will have to modify your code very slightly.

**Comment by kukreja-vikramaditya** [comment]
Please view wrecker failure..I didn't change anything from previous commit apart from test config method usage

**Comment by rnveach** [comment]
@kukreja-vikramaditya wercker failure is failure in master from one of our files in our another repo.
It can be ignored. We will rerun wercker once issue is fixed.
